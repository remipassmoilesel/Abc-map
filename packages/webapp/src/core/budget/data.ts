/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { I18nText, Language } from '@abc-map/shared';
import { BudgetItem } from './BudgetItem';

export const Expenses: BudgetItem[] = [
  {
    id: '2020-1',
    description: hostingLabel(2020),
    value: -633.6,
  },
  {
    id: '2021-1',
    description: hostingLabel(2021),
    value: -633.6,
  },
  {
    id: '2021-2',
    description: [
      { language: Language.French, text: 'Prestation UX 2021' },
      { language: Language.English, text: 'UX service 2021' },
    ],
    value: -95.05,
  },
  {
    id: '2022-1',
    description: hostingLabel(2022),
    value: -633.6,
  },
  {
    id: '2023-1',
    description: hostingLabel(2023),
    value: -633.6,
  },
];

export const Income: BudgetItem[] = [
  {
    id: '2020-1',
    description: donationsLabel(2021),
    value: 85.59,
  },
  {
    id: '2021-1',
    description: donationsLabel(2021),
    value: 88.94,
  },
  {
    id: '2022-1',
    description: donationsLabel(2022),
    value: 9.16,
  },
  {
    id: '2023-1',
    description: donationsLabel(2023),
    value: 9.36,
  },
];

function donationsLabel(year: number): I18nText[] {
  return [
    { language: Language.French, text: 'Dons ' + year },
    { language: Language.English, text: 'Donations ' + year },
  ];
}

function hostingLabel(year: number): I18nText[] {
  return [
    { language: Language.French, text: 'Hébergement ' + year },
    { language: Language.English, text: 'Hosting ' + year },
  ];
}
