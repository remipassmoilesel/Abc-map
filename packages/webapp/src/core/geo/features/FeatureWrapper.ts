/**
 * Copyright © 2021 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import Feature from 'ol/Feature';
import Geometry from 'ol/geom/Geometry';
import { AbcGeometryType, DefaultStyle, FeatureProperties, FeatureStyle, Logger, StyleProperties } from '@abc-map/shared';
import { nanoid } from 'nanoid';
import { OlGeometry } from './OlGeometry';
import { isOpenlayersFeature } from '../../utils/crossContextInstanceof';
import { GeoJSON } from 'ol/format';
import { Projection } from 'ol/proj';
import { GeoJSONFeature } from 'ol/format/GeoJSON';
import { DataRow } from '../../data/data-source/DataSource';
import { isValidDataField } from './isValidDataField';
import { HighlightedStyleFactory } from '../styles/HighlightedStyleFactory';

const logger = Logger.get('FeatureWrapper.ts');

export function disableFeatureWrapperLogging() {
  logger.disable();
}

export declare type OlPropertiesMap = { [key: string]: any };

export declare type DataPropertiesMap = { [key: string]: string | boolean | number | undefined | null };

const geoJson = new GeoJSON();

const highLightStyleFactory = new HighlightedStyleFactory();

export const HlPreviousStyleKey = 'abc:highlight:previous-style';
export const HlStyleKey = 'abc:highlight:highlight-style';

export type SupportedFeature<T extends OlGeometry = OlGeometry> = FeatureWrapper | Feature<T>;

/**
 * This class is a thin wrapper around Openlayers features, used to ensure that critical operations
 * on features are well done
 */
export class FeatureWrapper<Geom extends OlGeometry = OlGeometry> {
  public static create(geom?: Geometry): FeatureWrapper {
    return new FeatureWrapper(new Feature<Geometry>(geom)).setId();
  }

  public static from<T extends OlGeometry = OlGeometry>(feature: SupportedFeature): FeatureWrapper<T> {
    // If feature is already wrapped, we return it
    if (isFeatureWrapper(feature)) {
      return feature as FeatureWrapper<T>;
    }

    if (!isOpenlayersFeature(feature)) {
      throw new Error(`Invalid feature: ${feature ?? 'undefined'}`);
    }

    return new FeatureWrapper<T>(feature as Feature<T>);
  }

  public static fromUnknown<T extends OlGeometry = OlGeometry>(feature: unknown): FeatureWrapper<T> | undefined {
    // If feature is already wrapped, we return it
    if (isFeatureWrapper(feature)) {
      return feature as FeatureWrapper<T>;
    }

    // This is not an Openlayer feature, we can not process it
    if (!isOpenlayersFeature(feature)) {
      return;
    }

    return FeatureWrapper.from(feature as Feature<T>);
  }

  public static fromGeoJSON(feature: GeoJSONFeature, dataProjection?: Projection | string): FeatureWrapper {
    return FeatureWrapper.from(geoJson.readFeature(feature, { dataProjection }));
  }

  constructor(private feature: Feature<Geom>) {}

  public unwrap(): Feature<Geom> {
    return this.feature;
  }

  /**
   * Set id of feature if specified, or generate one
   *
   * @param id
   */
  public setId(id?: string | number): FeatureWrapper {
    this.feature.setId(id ?? nanoid(10));
    return this;
  }

  public getId(): string | number | undefined {
    return this.feature.getId();
  }

  /**
   * Clone feature, including its id
   */
  public clone(): FeatureWrapper {
    const clone = this.feature.clone();
    clone.setId(this.feature.getId());
    return FeatureWrapper.from(clone);
  }

  public isSelected(): boolean {
    return !!this.feature.get(FeatureProperties.Selected);
  }

  /**
   * WARNING: You should probably use FeatureSelection class instead.
   * @param value
   */
  public setSelected(value: boolean): FeatureWrapper {
    this.feature.set(FeatureProperties.Selected, value);
    return this;
  }

  /**
   * Extract style properties from feature
   */
  public getStyleProperties(): FeatureStyle {
    return {
      stroke: {
        color: this.feature.get(StyleProperties.StrokeColor),
        width: this.feature.get(StyleProperties.StrokeWidth),
      },
      fill: {
        color1: this.feature.get(StyleProperties.FillColor1),
        color2: this.feature.get(StyleProperties.FillColor2),
        pattern: this.feature.get(StyleProperties.FillPattern),
      },
      text: {
        value: this.feature.get(StyleProperties.TextValue),
        color: this.feature.get(StyleProperties.TextColor),
        size: this.feature.get(StyleProperties.TextSize),
        font: this.feature.get(StyleProperties.TextFont),
        offsetX: this.feature.get(StyleProperties.TextOffsetX),
        offsetY: this.feature.get(StyleProperties.TextOffsetY),
        rotation: this.feature.get(StyleProperties.TextRotation),
        alignment: this.feature.get(StyleProperties.TextAlignment),
      },
      point: {
        icon: this.feature.get(StyleProperties.PointIcon),
        size: this.feature.get(StyleProperties.PointSize),
        color: this.feature.get(StyleProperties.PointColor),
      },
      zIndex: this.feature.get(StyleProperties.ZIndex),
    };
  }

  /**
   * Set style properties on feature.
   *
   * Properties are set only if they are defined.
   *
   * @param properties
   */
  public setStyleProperties(properties: FeatureStyle): FeatureWrapper {
    const setIfDefined = (name: string, value: string | number | undefined) => {
      if (typeof value !== 'undefined') {
        this.feature.set(name, value);
      }
    };

    setIfDefined(StyleProperties.StrokeColor, properties.stroke?.color);
    setIfDefined(StyleProperties.StrokeWidth, properties.stroke?.width);
    setIfDefined(StyleProperties.FillColor1, properties.fill?.color1);
    setIfDefined(StyleProperties.FillColor2, properties.fill?.color2);
    setIfDefined(StyleProperties.FillPattern, properties.fill?.pattern);
    setIfDefined(StyleProperties.TextValue, properties.text?.value);
    setIfDefined(StyleProperties.TextColor, properties.text?.color);
    setIfDefined(StyleProperties.TextSize, properties.text?.size);
    setIfDefined(StyleProperties.TextFont, properties.text?.font);
    setIfDefined(StyleProperties.TextOffsetX, properties.text?.offsetX);
    setIfDefined(StyleProperties.TextOffsetY, properties.text?.offsetY);
    setIfDefined(StyleProperties.TextRotation, properties.text?.rotation);
    setIfDefined(StyleProperties.TextAlignment, properties.text?.alignment);
    setIfDefined(StyleProperties.PointIcon, properties.point?.icon);
    setIfDefined(StyleProperties.PointSize, properties.point?.size);
    setIfDefined(StyleProperties.PointColor, properties.point?.color);
    setIfDefined(StyleProperties.ZIndex, properties.zIndex);

    return this;
  }

  public setDefaultStyle(): FeatureWrapper {
    const geom = this.getGeometry();
    if (!geom) {
      return this;
    }

    switch (geom.getType()) {
      case AbcGeometryType.POLYGON:
      case AbcGeometryType.MULTI_POLYGON:
      case AbcGeometryType.CIRCLE:
      case AbcGeometryType.GEOMETRY_COLLECTION:
        this.setStyleProperties({ fill: DefaultStyle.fill, stroke: DefaultStyle.stroke });
        break;
      case AbcGeometryType.POINT:
      case AbcGeometryType.MULTI_POINT:
        this.setStyleProperties({ point: DefaultStyle.point });
        break;
      case AbcGeometryType.LINE_STRING:
      case AbcGeometryType.MULTI_LINE_STRING:
      case AbcGeometryType.LINEAR_RING:
        this.setStyleProperties({ stroke: DefaultStyle.stroke });
        break;
    }

    if (this.getText()) {
      this.setStyleProperties({ text: DefaultStyle.text });
    }

    return this;
  }

  public getGeometry(): Geom | undefined {
    return this.feature.getGeometry();
  }

  /**
   * If you do not specify geometry types, returns true if there is a geometry attached.
   *
   * If you do specify geometry types, returns true if there is a geometry with specified type, false otherwise.
   *
   * @param geoms
   */
  public hasGeometry(...geoms: AbcGeometryType[]): boolean {
    if (!geoms.length) {
      return !!this.getGeometry();
    }

    const type = this.getGeometry()?.getType();
    return !!geoms.find((t) => t === type);
  }

  public setText(text: string): FeatureWrapper {
    const style = this.getStyleProperties();
    this.setStyleProperties({
      ...style,
      text: {
        ...style.text,
        value: text,
      },
    });
    return this;
  }

  public getText(): string | undefined {
    return this.getStyleProperties().text?.value;
  }

  public getAllProperties(): OlPropertiesMap {
    return this.feature.getProperties();
  }

  /**
   * Return properties of feature:
   * - that are not internal
   * - that are numbers or strings
   */
  public getDataProperties(): DataPropertiesMap {
    const result: DataPropertiesMap = {};
    const properties = this.feature.getProperties();

    for (const key in properties) {
      const property = properties[key];
      const typeIsCorrect = isValidDataField(property);
      const notAbcProperty = key.indexOf('abc:') === -1;
      if (typeIsCorrect && notAbcProperty) {
        result[key] = property;
      }
    }

    return result;
  }

  public setProperties(properties: DataPropertiesMap): FeatureWrapper {
    this.feature.setProperties(properties);
    return this;
  }

  /**
   * Overwrite all data properties.
   *
   * @param properties
   */
  public setDataProperties(properties: DataPropertiesMap): FeatureWrapper {
    const original = this.getDataProperties();
    const propertyKeys = Object.keys(properties);

    // First we remove properties
    const toRemove = Object.keys(original).filter((k) => !propertyKeys.find((k2) => k === k2));
    toRemove.forEach((k) => this.feature.unset(k));

    // Then we set and trigger change
    this.feature.setProperties(properties);
    return this;
  }

  public toGeoJSON(dataProjection?: Projection | string): GeoJSONFeature {
    const clone = this.clone();
    const data = clone.getAllProperties();

    // We must delete non serializable properties
    for (const key in data) {
      const field = data[key];

      if (key !== 'geometry' && !isValidDataField(field)) {
        clone.unwrap().unset(key);
      }
    }

    return geoJson.writeFeatureObject(clone.unwrap(), { dataProjection });
  }

  public toDataRow(): DataRow {
    let id = this.getId();
    if (!id) {
      logger.warn('Feature does not have an id, we generate a new one');
      this.setId();
      id = this.getId();
      if (!id) {
        throw new Error('Invalid id');
      }
    }

    return {
      id,
      metadata: {
        selected: this.isSelected(),
      },
      data: this.getDataProperties(),
    };
  }

  public isHighlighted(): boolean {
    return !!this.feature.get(HlStyleKey);
  }

  public setHighlighted(value: boolean, timeoutMs = -1): void {
    if (this.isHighlighted() === value) {
      return;
    }

    if (value) {
      this.feature.set(HlPreviousStyleKey, this.feature.getStyle());

      const style = highLightStyleFactory.getForGeometry(this.getGeometry(), this.getStyleProperties());
      this.feature.setStyle(style);

      this.feature.set(HlStyleKey, style);

      if (timeoutMs > 0) {
        setTimeout(() => this.setHighlighted(false), timeoutMs);
      }
    } else {
      const previousStyle = this.feature.get(HlPreviousStyleKey);
      this.feature.setStyle(previousStyle);

      this.feature.unset(HlPreviousStyleKey);
      this.feature.unset(HlStyleKey);
    }
  }

  /**
   * This function helps to identify a FeatureWrapper, even in case of cross context objects.
   *
   * In most cases instanceof may work, but not for modules.
   */
  public isFeatureWrapper(): this is FeatureWrapper<Geom> {
    return true;
  }
}

export function isFeatureWrapper(feature: unknown) {
  return (
    !!feature && typeof feature === 'object' && 'isFeatureWrapper' in feature && typeof feature.isFeatureWrapper === 'function' && feature.isFeatureWrapper()
  );
}
