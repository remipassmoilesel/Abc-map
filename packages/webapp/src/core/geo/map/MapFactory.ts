/**
 * Copyright © 2021 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import Map from 'ol/Map';
import { MapWrapper } from './MapWrapper';
import { Views } from '../Views';

export class MapFactory {
  public static createDefault(empty = false): MapWrapper {
    const internal = new Map({
      layers: [],
      controls: [],
      view: Views.defaultOlView(),
    });

    const map = MapWrapper.from(internal);
    !empty && map.setDefaultLayers();
    map.setDefaultTool();

    return map;
  }

  public static createLayoutPreview(): MapWrapper {
    const internal = new Map({
      controls: [],
      interactions: [],
      layers: [],
      view: Views.defaultOlView(),
    });

    const map = MapWrapper.from(internal);
    map.setDefaultTool();

    return map;
  }

  public static createNaked(): MapWrapper {
    const internal = new Map({
      controls: [],
      interactions: [],
      layers: [],
      view: Views.defaultOlView(),
    });

    return MapWrapper.from(internal);
  }
}
