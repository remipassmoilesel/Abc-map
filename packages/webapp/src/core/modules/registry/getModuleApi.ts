/**
 * Copyright © 2021 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { LayerWrapper as ModuleLayerWrapper, FeatureWrapper as ModuleFeatureWrapper, ModuleApi } from '@abc-map/module-api';
import { getServices } from '../../../core/Services';
import { FeatureWrapper } from '../../../core/geo/features/FeatureWrapper';
import { LayerFactory } from '../../../core/geo/layers/LayerFactory';
import { MapWrapper } from '../../../core/geo/map/MapWrapper';
import { AddLayersChangeset } from '../../../core/history/changesets/layers/AddLayersChangeset';
import { RemoveLayerChangeset } from '../../../core/history/changesets/layers/RemoveLayerChangeset';
import { AddFeaturesChangeset } from '../../../core/history/changesets/features/AddFeaturesChangeset';
import { LayerWrapper } from '../../../core/geo/layers/LayerWrapper';
import VectorSource from 'ol/source/Vector';
import { RemoveFeaturesChangeset } from '../../../core/history/changesets/features/RemoveFeaturesChangeset';

/**
 * This API is used in external modules and in scripts
 */
export function getModuleApi(): ModuleApi {
  const { geo, history, modals } = getServices();

  return {
    FeatureWrapperFactory: {
      from: FeatureWrapper.from,
      fromUnknown: FeatureWrapper.fromUnknown,
    },
    LayerFactory: LayerFactory,
    MapFactory: {
      from: MapWrapper.from,
      fromUnknown: MapWrapper.fromUnknown,
    },
    changesets: {
      // Here we must cast entites because ModuleEntities are subset of entities
      addLayers: (layers: ModuleLayerWrapper[]) => AddLayersChangeset.create(layers as LayerWrapper[]),
      removeLayer: (layer: ModuleLayerWrapper) => RemoveLayerChangeset.create(layer as LayerWrapper),
      addFeatures: (source: VectorSource, features: ModuleFeatureWrapper[]) => new AddFeaturesChangeset(source, features as FeatureWrapper[]),
      removeFeatures: (source: VectorSource, features: ModuleFeatureWrapper[]) => new RemoveFeaturesChangeset(source, features as FeatureWrapper[]),
    },
    mainMap: geo.getMainMap(),
    services: {
      geo,
      history,
      modals,
    },
    // This variable will be set at runtime
    resourceBaseUrl: '',
  };
}
