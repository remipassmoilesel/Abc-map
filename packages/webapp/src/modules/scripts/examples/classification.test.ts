/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { MapWrapper } from '../../../core/geo/map/MapWrapper';
import { Classification } from './index';
import { MapFactory } from '../../../core/geo/map/MapFactory';
import sinon, { SinonStub } from 'sinon';
import * as fs from 'fs';
import * as path from 'path';
import { createScript } from '../script-api/createScript';
import { ScriptContext } from '../script-api/ScriptContext';
import { getModuleApi } from '../../../core/modules/registry/getModuleApi';
import { VectorLayerWrapper } from '../../../core/geo/layers/LayerWrapper';
import { newTestServices, TestServices } from '../../../core/utils/test/TestServices';
import { FeatureWrapper } from '../../../core/geo/features/FeatureWrapper';
import { getScriptApi } from '../script-api/ScriptApi';
import sortBy from 'lodash/sortBy';
import { logger } from './data';

logger.disable();

describe('classification.js-txt', () => {
  let logStub: sinon.SinonStub;
  let map: MapWrapper;
  let testServices: TestServices;

  beforeEach(async () => {
    logStub = sinon.stub();
    map = MapFactory.createDefault();

    testServices = newTestServices();
    testServices.geo.getMainMap.returns(map);

    (fetch as SinonStub).callsFake(async function (url: string | undefined) {
      if (url?.endsWith('.geojson')) {
        const filePath = path.resolve(__dirname, 'data/' + url);
        return fs.promises.readFile(filePath).then((res) => ({ json: async () => res.toString('utf-8') }));
      }
    });
  });

  it('setup should work', async () => {
    await Classification.setup(map);

    const layerNames = map.getLayers().map((layer) => layer.getName());
    expect(layerNames).toMatchSnapshot();
  });

  it('example should work', async () => {
    // Prepare
    await Classification.setup(map);

    const scriptContent = await fs.promises.readFile(path.resolve(__dirname, Classification.codeUrl));
    const script = createScript(scriptContent.toString('utf-8'));

    const context: ScriptContext = {
      moduleApi: getModuleApi(testServices),
      scriptApi: getScriptApi(testServices),
      log: logStub,
    };

    // Act
    await script(context);

    // Assert
    const layers = map.getLayers();
    const updatedLayer = layers[layers.length - 1] as VectorLayerWrapper;
    expect(updatedLayer.isVector()).toBe(true);

    const featureStyles = updatedLayer
      .getSource()
      .getFeatures()
      .map((feature) => FeatureWrapper.from(feature).getStyleProperties());
    expect(sortBy(featureStyles, (v) => v.text?.value)).toMatchSnapshot();
  });
});
