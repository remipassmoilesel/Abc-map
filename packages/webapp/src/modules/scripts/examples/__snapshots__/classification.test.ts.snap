// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`classification.js-txt example should work 1`] = `
Array [
  Object {
    "fill": Object {
      "color1": "#006abd",
      "color2": "rgba(255,255,255,0.60)",
      "pattern": "abc:style:fill:hatching:oblique:right",
    },
    "point": Object {
      "color": undefined,
      "icon": undefined,
      "size": undefined,
    },
    "stroke": Object {
      "color": "rgba(18,90,147,0.60)",
      "width": 3,
    },
    "text": Object {
      "alignment": undefined,
      "color": undefined,
      "font": undefined,
      "offsetX": undefined,
      "offsetY": undefined,
      "rotation": undefined,
      "size": 15,
      "value": undefined,
    },
    "zIndex": undefined,
  },
  Object {
    "fill": Object {
      "color1": "#006abd",
      "color2": "rgba(255,255,255,0.60)",
      "pattern": "abc:style:fill:hatching:oblique:right",
    },
    "point": Object {
      "color": undefined,
      "icon": undefined,
      "size": undefined,
    },
    "stroke": Object {
      "color": "rgba(18,90,147,0.60)",
      "width": 3,
    },
    "text": Object {
      "alignment": undefined,
      "color": undefined,
      "font": undefined,
      "offsetX": undefined,
      "offsetY": undefined,
      "rotation": undefined,
      "size": undefined,
      "value": undefined,
    },
    "zIndex": undefined,
  },
  Object {
    "fill": Object {
      "color1": "#9fd6ff",
      "color2": "rgba(255,255,255,0.60)",
      "pattern": "abc:style:fill:hatching:oblique:right",
    },
    "point": Object {
      "color": undefined,
      "icon": undefined,
      "size": undefined,
    },
    "stroke": Object {
      "color": "rgba(18,90,147,0.60)",
      "width": 3,
    },
    "text": Object {
      "alignment": undefined,
      "color": undefined,
      "font": undefined,
      "offsetX": undefined,
      "offsetY": undefined,
      "rotation": undefined,
      "size": undefined,
      "value": undefined,
    },
    "zIndex": undefined,
  },
  Object {
    "fill": Object {
      "color1": "#006abd",
      "color2": "rgba(255,255,255,0.60)",
      "pattern": "abc:style:fill:hatching:oblique:right",
    },
    "point": Object {
      "color": undefined,
      "icon": undefined,
      "size": undefined,
    },
    "stroke": Object {
      "color": "rgba(18,90,147,0.60)",
      "width": 3,
    },
    "text": Object {
      "alignment": undefined,
      "color": undefined,
      "font": undefined,
      "offsetX": undefined,
      "offsetY": undefined,
      "rotation": undefined,
      "size": undefined,
      "value": undefined,
    },
    "zIndex": undefined,
  },
  Object {
    "fill": Object {
      "color1": "#9fd6ff",
      "color2": "rgba(255,255,255,0.60)",
      "pattern": "abc:style:fill:hatching:oblique:right",
    },
    "point": Object {
      "color": undefined,
      "icon": undefined,
      "size": undefined,
    },
    "stroke": Object {
      "color": "rgba(18,90,147,0.60)",
      "width": 3,
    },
    "text": Object {
      "alignment": undefined,
      "color": undefined,
      "font": undefined,
      "offsetX": undefined,
      "offsetY": undefined,
      "rotation": undefined,
      "size": undefined,
      "value": undefined,
    },
    "zIndex": undefined,
  },
  Object {
    "fill": Object {
      "color1": "#006abd",
      "color2": "rgba(255,255,255,0.60)",
      "pattern": "abc:style:fill:hatching:oblique:right",
    },
    "point": Object {
      "color": undefined,
      "icon": undefined,
      "size": undefined,
    },
    "stroke": Object {
      "color": "rgba(18,90,147,0.60)",
      "width": 3,
    },
    "text": Object {
      "alignment": undefined,
      "color": undefined,
      "font": undefined,
      "offsetX": undefined,
      "offsetY": undefined,
      "rotation": undefined,
      "size": undefined,
      "value": undefined,
    },
    "zIndex": undefined,
  },
  Object {
    "fill": Object {
      "color1": "#006abd",
      "color2": "rgba(255,255,255,0.60)",
      "pattern": "abc:style:fill:hatching:oblique:right",
    },
    "point": Object {
      "color": undefined,
      "icon": undefined,
      "size": undefined,
    },
    "stroke": Object {
      "color": "rgba(18,90,147,0.60)",
      "width": 3,
    },
    "text": Object {
      "alignment": undefined,
      "color": undefined,
      "font": undefined,
      "offsetX": undefined,
      "offsetY": undefined,
      "rotation": undefined,
      "size": undefined,
      "value": undefined,
    },
    "zIndex": undefined,
  },
  Object {
    "fill": Object {
      "color1": "#00447c",
      "color2": "rgba(255,255,255,0.60)",
      "pattern": "abc:style:fill:hatching:oblique:right",
    },
    "point": Object {
      "color": undefined,
      "icon": undefined,
      "size": undefined,
    },
    "stroke": Object {
      "color": "rgba(18,90,147,0.60)",
      "width": 3,
    },
    "text": Object {
      "alignment": undefined,
      "color": undefined,
      "font": undefined,
      "offsetX": undefined,
      "offsetY": undefined,
      "rotation": undefined,
      "size": undefined,
      "value": undefined,
    },
    "zIndex": undefined,
  },
  Object {
    "fill": Object {
      "color1": "#9fd6ff",
      "color2": "rgba(255,255,255,0.60)",
      "pattern": "abc:style:fill:hatching:oblique:right",
    },
    "point": Object {
      "color": undefined,
      "icon": undefined,
      "size": undefined,
    },
    "stroke": Object {
      "color": "rgba(18,90,147,0.60)",
      "width": 3,
    },
    "text": Object {
      "alignment": undefined,
      "color": undefined,
      "font": undefined,
      "offsetX": undefined,
      "offsetY": undefined,
      "rotation": undefined,
      "size": undefined,
      "value": undefined,
    },
    "zIndex": undefined,
  },
  Object {
    "fill": Object {
      "color1": "#00447c",
      "color2": "rgba(255,255,255,0.60)",
      "pattern": "abc:style:fill:hatching:oblique:right",
    },
    "point": Object {
      "color": undefined,
      "icon": undefined,
      "size": undefined,
    },
    "stroke": Object {
      "color": "rgba(18,90,147,0.60)",
      "width": 3,
    },
    "text": Object {
      "alignment": undefined,
      "color": undefined,
      "font": undefined,
      "offsetX": undefined,
      "offsetY": undefined,
      "rotation": undefined,
      "size": undefined,
      "value": undefined,
    },
    "zIndex": undefined,
  },
  Object {
    "fill": Object {
      "color1": "#9fd6ff",
      "color2": "rgba(255,255,255,0.60)",
      "pattern": "abc:style:fill:hatching:oblique:right",
    },
    "point": Object {
      "color": undefined,
      "icon": undefined,
      "size": undefined,
    },
    "stroke": Object {
      "color": "rgba(18,90,147,0.60)",
      "width": 3,
    },
    "text": Object {
      "alignment": undefined,
      "color": undefined,
      "font": undefined,
      "offsetX": undefined,
      "offsetY": undefined,
      "rotation": undefined,
      "size": undefined,
      "value": undefined,
    },
    "zIndex": undefined,
  },
  Object {
    "fill": Object {
      "color1": "#00447c",
      "color2": "rgba(255,255,255,0.60)",
      "pattern": "abc:style:fill:hatching:oblique:right",
    },
    "point": Object {
      "color": undefined,
      "icon": undefined,
      "size": undefined,
    },
    "stroke": Object {
      "color": "rgba(18,90,147,0.60)",
      "width": 3,
    },
    "text": Object {
      "alignment": undefined,
      "color": undefined,
      "font": undefined,
      "offsetX": undefined,
      "offsetY": undefined,
      "rotation": undefined,
      "size": undefined,
      "value": undefined,
    },
    "zIndex": undefined,
  },
  Object {
    "fill": Object {
      "color1": "#00447c",
      "color2": "rgba(255,255,255,0.60)",
      "pattern": "abc:style:fill:hatching:oblique:right",
    },
    "point": Object {
      "color": undefined,
      "icon": undefined,
      "size": undefined,
    },
    "stroke": Object {
      "color": "rgba(18,90,147,0.60)",
      "width": 3,
    },
    "text": Object {
      "alignment": undefined,
      "color": undefined,
      "font": undefined,
      "offsetX": undefined,
      "offsetY": undefined,
      "rotation": undefined,
      "size": undefined,
      "value": undefined,
    },
    "zIndex": undefined,
  },
  Object {
    "fill": Object {
      "color1": "#006abd",
      "color2": "rgba(255,255,255,0.60)",
      "pattern": "abc:style:fill:hatching:oblique:right",
    },
    "point": Object {
      "color": undefined,
      "icon": undefined,
      "size": undefined,
    },
    "stroke": Object {
      "color": "rgba(18,90,147,0.60)",
      "width": 3,
    },
    "text": Object {
      "alignment": undefined,
      "color": undefined,
      "font": undefined,
      "offsetX": undefined,
      "offsetY": undefined,
      "rotation": undefined,
      "size": undefined,
      "value": undefined,
    },
    "zIndex": undefined,
  },
  Object {
    "fill": Object {
      "color1": "#006abd",
      "color2": "rgba(255,255,255,0.60)",
      "pattern": "abc:style:fill:hatching:oblique:right",
    },
    "point": Object {
      "color": undefined,
      "icon": undefined,
      "size": undefined,
    },
    "stroke": Object {
      "color": "rgba(18,90,147,0.60)",
      "width": 3,
    },
    "text": Object {
      "alignment": undefined,
      "color": undefined,
      "font": undefined,
      "offsetX": undefined,
      "offsetY": undefined,
      "rotation": undefined,
      "size": 15,
      "value": undefined,
    },
    "zIndex": undefined,
  },
  Object {
    "fill": Object {
      "color1": "#9fd6ff",
      "color2": "rgba(255,255,255,0.60)",
      "pattern": "abc:style:fill:hatching:oblique:right",
    },
    "point": Object {
      "color": undefined,
      "icon": undefined,
      "size": undefined,
    },
    "stroke": Object {
      "color": "rgba(18,90,147,0.60)",
      "width": 3,
    },
    "text": Object {
      "alignment": undefined,
      "color": undefined,
      "font": undefined,
      "offsetX": undefined,
      "offsetY": undefined,
      "rotation": undefined,
      "size": 15,
      "value": undefined,
    },
    "zIndex": undefined,
  },
  Object {
    "fill": Object {
      "color1": "#9fd6ff",
      "color2": "rgba(255,255,255,0.60)",
      "pattern": "abc:style:fill:hatching:oblique:right",
    },
    "point": Object {
      "color": undefined,
      "icon": undefined,
      "size": undefined,
    },
    "stroke": Object {
      "color": "rgba(18,90,147,0.60)",
      "width": 3,
    },
    "text": Object {
      "alignment": undefined,
      "color": undefined,
      "font": undefined,
      "offsetX": undefined,
      "offsetY": undefined,
      "rotation": undefined,
      "size": undefined,
      "value": undefined,
    },
    "zIndex": undefined,
  },
  Object {
    "fill": Object {
      "color1": "#9fd6ff",
      "color2": "rgba(255,255,255,0.60)",
      "pattern": "abc:style:fill:hatching:oblique:right",
    },
    "point": Object {
      "color": undefined,
      "icon": undefined,
      "size": undefined,
    },
    "stroke": Object {
      "color": "rgba(18,90,147,0.60)",
      "width": 3,
    },
    "text": Object {
      "alignment": undefined,
      "color": undefined,
      "font": undefined,
      "offsetX": undefined,
      "offsetY": undefined,
      "rotation": undefined,
      "size": undefined,
      "value": undefined,
    },
    "zIndex": undefined,
  },
]
`;

exports[`classification.js-txt setup should work 1`] = `
Array [
  "OpenStreetMap",
  "Geometries",
  "Regions of France",
]
`;
